src/catalog/src/expr_cache.rs

```diff
@@ -0,0 +1,592 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A cache for optimized expressions.
+
+#![allow(dead_code)]
+
+use std::borrow::Cow;
+use std::collections::{BTreeMap, BTreeSet};
+use std::future::Future;
+use std::sync::Arc;
+
+use mz_compute_types::dataflows::{DataflowDescription, DataflowExpirationDesc};
+use mz_durable_cache::{DurableCache, DurableCacheCodec};
+use mz_expr::{MirRelationExpr, OptimizedMirRelationExpr};
+use mz_ore::channel::trigger;
+use mz_ore::task::spawn;
+use mz_persist_client::PersistClient;
+use mz_persist_types::codec_impls::UnitSchema;
+use mz_persist_types::Codec;
+use mz_repr::adt::jsonb::Jsonb;
+use mz_repr::optimize::OptimizerFeatures;
+use mz_repr::{GlobalId, RelationDesc, RelationType, ScalarType};
+use mz_storage_types::sources::SourceData;
+use mz_transform::dataflow::DataflowMetainfo;
+use mz_transform::notice::OptimizerNotice;
+use proptest::arbitrary::{any, Arbitrary};
+use proptest::prelude::BoxedStrategy;
+use proptest::strategy::Strategy;
+use proptest_derive::Arbitrary;
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+use timely::progress::Antichain;
+use timely::Container;
+use tokio::sync::mpsc;
+use tracing::debug;
+use uuid::Uuid;
+
+use crate::durable::expression_cache_shard_id;
+
+/// The data that is cached per catalog object.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Expressions {
+    local_mir: OptimizedMirRelationExpr,
+    global_mir: DataflowDescription<OptimizedMirRelationExpr>,
+    physical_plan: DataflowDescription<mz_compute_types::plan::Plan>,
+    dataflow_metainfos: DataflowMetainfo<Arc<OptimizerNotice>>,
+    notices: SmallVec<[Arc<OptimizerNotice>; 4]>,
+    optimizer_features: OptimizerFeatures,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Arbitrary)]
+struct CacheKey {
+    deploy_generation: u64,
+    id: GlobalId,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct CodecKey<'a> {
+    cache_key: Cow<'a, CacheKey>,
+    expressions: Cow<'a, Expressions>,
+}
+
+#[derive(Debug)]
+struct ExpressionCodec;
+
+impl DurableCacheCodec for ExpressionCodec {
+    type Key = CacheKey;
+    type Val = Expressions;
+    type KeyCodec = SourceData;
+    type ValCodec = ();
+
+    fn schemas() -> (
+        <Self::KeyCodec as Codec>::Schema,
+        <Self::ValCodec as Codec>::Schema,
+    ) {
+        (
+            RelationDesc::builder()
+                .with_column("data", ScalarType::Jsonb.nullable(false))
```

Comment:
I've been thinking since we initially talked about all this and have slightly warmed up to two fields instead: "key" and "val", both `jsonb`. It would be harder to change away from the key/val structure, but I don't want to overindex on my regrets for txn-wal and how hard it was to add in the schema later. To be clear, I'm on the fence and either is probably fine but...

The advantage to two fields:
- More semantic
- Perhaps better persist pushdown stats, they're recursively computed and trimmed, but ultimately stored per-field
- Maybe something something future projection pushdown handwave

Reply:
As in something like

```Rust
        (
            RelationDesc::builder()
                .with_column("key", ScalarType::Jsonb.nullable(false))
                .finish(),
            RelationDesc::builder()
                .with_column("val", ScalarType::Jsonb.nullable(false))
                .finish(),
        )
```

?

Another advantage of this approach that I'm just realizing is that we can more easily deserialize the key and value separately in case of backwards compatibility issues.

Reply:
Yeah, exactly

Reply:
I'm going to merge as is, but I might have to change this when I try and implement backwards compatibility in the next PR.

Reply:
The more I think about it, the more I think you might want separate fields for persist stats reasons. So it might be a good idea to switch before we roll this out too widely. BUT, unlike txn-wal, we could always easily burn `const EXPRESSION_CACHE_SEED: usize = 4;` so maybe it really is fine to punt until/if it becomes a problem